### PR TITLE
Upgrade to TS 4.0

### PR DESCRIPTION
**Pull request by avatarneil**
## 📚 Purpose
Upgrades us from TS 3.8 -> TS 4.0!

## 📦 Impacts:
Everything

### CODE REVIEW

**Comment by janedotx** [comment]
I thought there were going to be a lot of Jest migrations in this?

File: packages/mds-policy-author/tests/api.spec.ts [hunk 1]
```diff
@@ -188,7 +193,12 @@ describe('Tests app', () => {
 
     it('verifies cannot PUT invalid policy', async () => {
       const bad_policy_json: Policy = clone(POLICY_JSON_WITHOUT_PUBLISH_DATE)
+
+      /* eslint-reason intentionally bad test */
+      /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
+      // @ts-ignore
```
Comment (drtyh2o) [inline packages/mds-policy-author/tests/api.spec.ts:199]: 😱

**Review by drtyh2o — COMMENTED** [review]
I agree with @janedotx about the conversion to `jest` and removal of mocha/nyc/etc as well. It should be pretty straight forward since I don't think there is any mocking going on in this repo presently.

**Comment by janedotx** [comment]
I just ran this locally and all tests pass. How is this possible?

**Comment by drtyh2o** [comment]
> I just ran this locally and all tests pass. How is this possible?

They always did. The issue was never with upgrading mds-core to TS 4.0 ... it's always been with being able to subsequently mock mds-core modules after they are published and then imported elsewhere.

**Comment by avatarneil** [comment]
Do we really need full jest conversion to be a blocker? I was thinking it'd be nice to just migrate the things that needed to be migrated, and then put together a plan for migrating everything over the next few months...

**Comment by janedotx** [comment]
> > I just ran this locally and all tests pass. How is this possible?
> 
> They always did. The issue was never with upgrading mds-core to TS 4.0 ... it's always been with being able to subsequently mock mds-core modules after they are published and then imported elsewhere.

👍

**Comment by janedotx** [comment]
Yeah, I didn't realize that mocks just failed for methods imported from other packages. It makes sense that tests pass, since the mocks used in mds-core are where the tests exist in the same package as the mocked functions.

> Do we really need full jest conversion to be a blocker? I was thinking it'd be nice to just migrate the things that needed to be migrated, and then put together a plan for migrating everything over the next few months...

I think it's fine to go ahead as is, but I think we all have a lot on our plate and a full conversion to jest, when it's not necessary to get tests passing in mds-core, seems likely to fall off the radar.

File: packages/mds-policy-author/tests/api.spec.ts [hunk 2]
```diff
@@ -96,7 +96,12 @@ describe('Tests app', () => {
 
     it('tries to post invalid policy', done => {
       const bad_policy_json: Policy = clone(POLICY_JSON_WITHOUT_PUBLISH_DATE)
+
+      /* eslint-reason intentionally bad test */
+      /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
+      // @ts-ignore
```
Comment (janedotx) [inline packages/mds-policy-author/tests/api.spec.ts:102]: Why? This seems like a good case to test.

Comment (avatarneil) [reply packages/mds-policy-author/tests/api.spec.ts:102]: The test is fine, I meant it's intentionally testing a "bad data" case. Linter yells when mutating an object and deleting a non-optional property. The test is doing something that would never happen in a fully-typed context.

**Comment by avatarneil** [comment]
> I think it's fine to go ahead as is, but I think we all have a lot on our plate and a full conversion to jest, when it's not necessary to get tests passing in mds-core, seems likely to fall off the radar.

I'll make sure it doesn't fall off the radar, even if I have to do it all myself 😜

File: packages/mds-policy-author/tests/api.spec.ts (hunk 2 above)
Comment (janedotx) [reply packages/mds-policy-author/tests/api.spec.ts:102]: Ok, I do hate being that person, but I feel like given that Michael and I both had reactions to these commented out lines, a little bit more explanation is required.

Comment (avatarneil) [reply packages/mds-policy-author/tests/api.spec.ts:104]: Updated in [this commit](https://github.com/lacuna-tech/mds-core/pull/427/commits/129d48096d7c67b8900867db9c11daf648e9a4dd)

**Comment by janedotx** [comment]
> > I think it's fine to go ahead as is, but I think we all have a lot on our plate and a full conversion to jest, when it's not necessary to get tests passing in mds-core, seems likely to fall off the radar.
> 
> I'll make sure it doesn't fall off the radar, even if I have to do it all myself 😜

Well, that was fast.

**Review by janedotx — APPROVED** [review]

**Review by drtyh2o — APPROVED** [review]

**Pull request merged**